app/carpool/views.py

```diff
@@ -290,32 +288,27 @@ def _email_carpool_cancelled(carpool, reason, send=False):
 
     subject = 'Carpool session on {} cancelled'.format(carpool.leave_time)
 
-    # TODO: This should be an HTML template stored elsewhere
-    body = '''
-        Hello rider,
-
-        Unfortunately, the carpool session for leaving from {} at {} has been
-        cancelled.
-
-        The reason given for the cancellation was: {}.
-
-        Please reach out to {} in order to see if they're willing
-        to reschedule.
-    '''.format(
-            carpool.from_place,
-            carpool.leave_time,
-            reason,
-            driver.email)
-
-    if send:
+    for rider in riders:
+        _send_email(
+            'carpools/email/carpool_cancelled.html',
+            'carpools/email/carpool_cancelled.txt',
+            rider.email,
+            subject,
+            driver=driver,
+            rider=rider,
+            carpool=carpool,
+            reason=reason)
+
+
+def _send_email(html_template, text_template, recipient, subject, **kwargs):
+    body = render_template(text_template, **kwargs)
+    html = render_template(html_template, **kwargs)
+    msg = Message(
+        recipients=[recipient], body=body, html=html, subject=subject)
+    try:
         with mail.connect() as conn:
```

Comment:
this func should maybe take a list of riders (or a list of emails -- which can be easily gotten from a list of riders via `[r.email for r in riders]`) and FIRST connect to the email client, and then iterate over emails and format + send email for each one -- I suspect connecting to the client is an amount of overhead and we won't wanna do it once for every single recipient.

might also be worth it to wrap in two different try's, so if something goes wrong sending a single email we send the rest:
```try:
    with mail.connect() as conn:
        for rider in riders:
            try:
                make_email(rider)
                send_email()
            except:
                log("problem sending email!")
except:
    log("problem connecting to mail client")```

Reply:
Thanks for the code review, @maiamcc.  

When I was testing this, I found that exceptions thrown in send_email() inside context `with mail.connect() as conn:` were caught in the context manager and propagated up the stack.  So in your example, which I believe is a better way to write the code (and it's how I originally wrote it), the exception was not caught in the `try` block and "problem sending email" was never caught. I realized this was happening when, in the exception's stack trace, I saw __enter__ from the context manager returned by mail.connect().  Does this make sense, because I would sure like a reality check.

Reply:
Thanks for the review, @maiamcc!